### PR TITLE
ROX-11564 replay tests with alerts

### PR DESCRIPTION
**Pull request by lvalerom**
## Description

This PR aims to provided testing for sensor's re-sync behaviour using pre-recorded k8s events.

The events were recorded using the local-sensor tool.

Create the policies.json file containing the policies:
```
curl -k -u admin:$ROX_ADMIN_PASSWORD -H "Content-Type:application/json" https://$ROX_CENTRAL_ADDRESS/v1/policies > $tmpf
# and then
# for each id
curl -k -u admin:$ROX_ADMIN_PASSWORD -H "Content-Type:application/json" https://$ROX_CENTRAL_ADDRESS/v1/policies/$id >> $fname 
```

Create the trace.jsonl file containing the k8s events.
```
go run tools/local-sensor/main.go -record -record-out=trace.jsonl -resync=0s
```
Create the central-out.bin file containing sensor's output events.
```
go run tools/local-sensor/main.go -replay -replay-in=trace.jsonl -resync=10s -delay=0s -format=raw -central-out=central-out.bin -with-policies=policies.json
```

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Manually testing with:
```
LOGLEVEL=panic go test -race -count=1 github.com/stackrox/rox/sensor/tests/replay
```
Disabling the re-sync by changing resyncTime := 1 * time.Second to resyncTime := 0 * time.Second will make the test fail.

### CODE REVIEW

**Comment by lvalerom** [comment]
Part of this PR is addressed here #2483

**Comment by vikin91** [comment]
I see some rebase is needed here - please ping me for a review once the conflicts are resolved

**Comment by lvalerom** [comment]
> I see some rebase is needed here - please ping me for a review once the conflicts are resolved

Yeah. The idea is to merge #2483 first and then rebase this one.

File: sensor/debugger/k8s/fake.go [hunk 1]
```diff
@@ -216,6 +219,188 @@ func (f *FakeEventsManager) Init() {
 		cronJobKind:               &batchv1.CronJob{},
 		podKind:                   &corev1.Pod{},
 	}
+	f.resourceListMap = map[string]func(reflect.Value, func(string) error) error{
```
Comment (fredrb) [inline sensor/debugger/k8s/fake.go:222]: If I understand this code correctly, the only reason why we have this map is to call `GetName()` on each element on the `Items` slice. Since every resource list has the same interface (`.Items` + `.GetName()`), and we're only interested in the `string` value from the resource itself, I suggest that we make this generic. 

[Here's an example from playground](https://go.dev/play/p/90kOgXPFkWU). 

I suggest writing a single reflect-based function that iterates on `.Items` and calls `.GetName()` on each element:
```golang
func getNameGeneric(obj reflect.Value, f func(string) error) error {
	errorList := errorhelpers.NewErrorList("resource list function")
	items := obj.FieldByName("Items")
	if !items.IsValid() {
		errorList.AddError(errors.New("invalid kind: missing `Items` property"))
	} else {
		for i := 0; i < items.Len(); i++ {
			getNameFn := items.Index(i).MethodByName("GetName")
			returnVals := getNameFn.Call([]reflect.Value{})
			if err := f(returnVals[0].String()); err != nil {
				errorList.AddError(err)
			}
		}	
	}
	return errorList.ToError()
}
```

File: sensor/debugger/k8s/fake.go [hunk 2]
```diff
@@ -436,6 +621,68 @@ func (f *FakeEventsManager) waitOnMode(events []string) error {
 	return nil
 }
 
+// execFuncToAllResourcesOfKind executes a given function to all resources of a given kind
```
Comment (fredrb) [inline sensor/debugger/k8s/fake.go:624]: I understand that maybe the goal was to provide a callback to each resource object . However, only the name seems passed. Which is enough for removing namespaces. I suggest rewording this.
Additionally, you could already provide an `interface{}` object parameter to the `execFunc` callback. 

IMO this would read better:
```suggestion
// forEachResourceName iterates on every resource of given kind and call `callbackFn` for each resource object name
```
Or `forEachResource` if you decide on passing the entire object to the callback.

Comment (fredrb) [reply sensor/debugger/k8s/fake.go:624]: Keep in mind that if you decide to pass the entire object you could remove the`GetName()` reflection above. It would be the responsibility of the caller to cast the `interface{}` into whatever resource is being listed.

File: sensor/tests/replay/replay_test.go [hunk 3]
```diff
@@ -293,6 +323,23 @@ func readSensorOutputFile(fname string) ([]*central.MsgFromSensor, error) {
 	return msgs, nil
 }
 
+func getAlerts(messages []*central.MsgFromSensor) map[string]map[string]*storage.Alert {
+	events := make(map[string]map[string]*storage.Alert)
+	for _, msg := range messages {
+		event := msg.GetEvent()
+		if event.GetAlertResults() != nil {
+			if event.GetAlertResults().GetDeploymentId() != "" {
+				alerts := make(map[string]*storage.Alert)
```
Comment (fredrb) [inline sensor/tests/replay/replay_test.go:332]: Maybe add the map size if you know it ahead of time:
```golang
alertResults := event.GetAlertResults().GetAlerts()
alerts := make(map[string]*storage.Alert, len(alertResults))
for _, a := range alertResults {
       alerts[a.GetPolicy().GetId()] = a
}
```

**Review by fredrb — CHANGES_REQUESTED** [review]
Requesting changes due to the reflection suggestion. Let me know if I missed something or if there is a reason not to do it.

File: sensor/debugger/k8s/fake.go (hunk 1 above)
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Yeah. I changed your suggestion a little bit. Let me know what you think.

File: sensor/debugger/k8s/fake.go (hunk 2 above)
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Done. I've decided to pass the reflected value of the resource instead of the name. Let me know what you think.

File: sensor/tests/replay/replay_test.go (hunk 3 above)
Comment (lvalerom) [reply sensor/tests/replay/replay_test.go]: 👍

**Review by fredrb — APPROVED** [review]

File: sensor/debugger/k8s/fake.go [hunk 4]
```diff
@@ -436,6 +437,123 @@ func (f *FakeEventsManager) waitOnMode(events []string) error {
...
+	}
+	errInt := returnVals[len(returnVals)-1].Interface()
+	if errInt != nil {
+		return errInt.(error)
+	}
+
+	errorList := errorhelpers.NewErrorList("for each resource")
+	items := reflect.Indirect(returnVals[0]).FieldByName("Items")
+	if !items.IsValid() {
+		errorList.AddError(errors.New("invalid kind: missing `Items` property"))
```
Comment (vikin91) [inline sensor/debugger/k8s/fake.go:458]: Minor: we could return here and save the `else` block

File: sensor/debugger/k8s/fake.go [hunk 5]
```diff
@@ -436,6 +437,123 @@ func (f *FakeEventsManager) waitOnMode(events []string) error {
...
+	return errorList.ToError()
+}
+
+// getNameFromObjectMeta returns the name of the resource
+func getNameFromObjectMeta(obj reflect.Value) (string, error) {
+	objMeta := obj.FieldByName("ObjectMeta")
+	if !objMeta.IsValid() {
+		return "", errors.New("resource does not have the ObjectMeta field")
+	}
+	objMetaPtr := objMeta.Addr()
```
Comment (vikin91) [inline sensor/debugger/k8s/fake.go:476]: `objMetaPtr` is used only in the next line - maybe a candidate to remove?

File: sensor/debugger/k8s/fake.go [hunk 6]
```diff
@@ -436,6 +437,123 @@ func (f *FakeEventsManager) waitOnMode(events []string) error {
...
+			cl = reflect.ValueOf(clFunc(ns))
+			if err := f.forEachResource(cl, func(resource reflect.Value) error {
+				name, err := getNameFromObjectMeta(resource)
+				if err != nil {
+					return err
+				}
+				retVals := runOp(removeAction, cl, reflect.ValueOf(name))
+				if len(retVals) == 0 {
+					return fmt.Errorf("expected 1 or 2 values from %s. Received: %d", removeAction, len(retVals))
+				}
```
Comment (vikin91) [inline sensor/debugger/k8s/fake.go:543]: I guess `len(retVals)` cannot be 3 or more? If it can, then maybe we should change it to

```suggestion
				if len(retVals) < 1 || len(retVals) > 2 {
					return fmt.Errorf("expected 1 or 2 values from %s. Received: %d", removeAction, len(retVals))
				}
```

**Review by vikin91 — APPROVED** [review]
Looks good, I found just minor and fully optional improvements

File: sensor/debugger/k8s/fake.go (hunk 4 above)
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Done

File: sensor/debugger/k8s/fake.go (hunk 5 above)
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Done

File: sensor/debugger/k8s/fake.go (hunk 6 above)
Comment (lvalerom) [reply sensor/debugger/k8s/fake.go]: Done!
